### DISCUSSION TITLE
yarn upgrade-interactive and yarn patches

### DISCUSSION THREAD

**Discussion by darkbasic** [yarnpkg/berry#5336] [Q&A]
How am I supposed to update my deps while *keeping* the patches I was applying to them? I've recently upstreamed some changes into some libraries and I was curious how yarn would handle patches being upstreamed. Ideally it would keep all the patches that still apply and, noticing a patch is no longer necessary, drop it. Instead yarn `upgrade-interactive` just dropped ALL of the patches for a simple `^2.1.1` -> `^2.1.2` upgrade without even bothering to check if the previous patches would still apply or not.
This is very annoying and prompts me to switch to [patch-package](https://www.npmjs.com/package/patch-package) instead. Am I doing something wrong? What's the suggest workflow to handle upgrades with local patches?

**Comment by arcanis** [comment]
Nice nickname, I learned programming with Dark Basic a long time ago ... 🙂

You're not doing something wrong per se, it's just that `yarn upgrade-interactive` isn't (currently) integrated with the `patch:` protocol. It should be possible to implement that (PR welcome), although making it clean may be a bit more of a challenge.

**Reply by darkbasic** [reply]
@arcanis people who know what darkbasic was -let alone having used it- are rare today ... :)

It's funny that you talk about the `patch:` protocol, because I've just (in this exact moment) tried v4 and discovered about it. v3 used resolutions and patching a patched packaged resulted in this ugly mess:

```
"graphql-modules": "^2.1.1",

"resolutions": {
  "graphql-modules@^2.1.1": "patch:graphql-modules@patch%3Agraphql-modules@npm%253A2.1.1%23./.yarn/patches/graphql-modules-npm-2.1.1-3778b53b5f.patch%3A%3Aversion=2.1.1&hash=848910&locator=sports%2540workspace%253A.#./.yarn/patches/graphql-modules-patch-48bc65c306.patch"
  }
```

This is how it looks like now:
```
"graphql-modules": "patch:graphql-modules@npm%3A2.1.1#~/.yarn/patches/graphql-modules-npm-2.1.1-3778b53b5f.patch",
```

I like that I don't have to "patch the patch", but I don't like the fact that when I run `yarn patch graphql-modules` it doesn't automatically apply existing patches in the generated directory. An option to do so (possibly opt-out) would be appreciated.

With v4 `yarn upgrade-interactive` now won't honor ranges for patched packages anymore. While being good because it doesn't support automatic patching it makes upgrading even more annoying because that's one more thing you will have to do manually.

When I saw the new `yarn patch -u` update option I was relieved because it would make things so much easier (theoretically now for each patched package you just have to manually upgrade in range and run yarn patch -u package_name) but somehow doesn't work:

```
niko@talos2 ~/devel/beach/server $ git diff
diff --git a/package.json b/package.json
index 207f268..964b4f6 100644
--- a/package.json
+++ b/package.json
@@ -76,7 +76,7 @@
     "dotenv": "^16.0.3",
     "graphql": "^16.6.0",
     "graphql-import-node": "^0.0.5",
-    "graphql-modules": "patch:graphql-modules@npm%3A2.1.1#~/.yarn/patches/graphql-modules-npm-2.1.1-3778b53b5f.patch",
+    "graphql-modules": "^2.1.2",
     "graphql-scalars": "^1.20.4",
     "reflect-metadata": "^0.1.13",
     "roarr": "^7.14.7",
diff --git a/yarn.lock b/yarn.lock
index 566b5c4..3d51439 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -5456,9 +5456,9 @@ __metadata:
   languageName: node
   linkType: hard
 
-"graphql-modules@npm:^2.1.1":
-  version: 2.1.1
-  resolution: "graphql-modules@npm:2.1.1"
+"graphql-modules@npm:^2.1.2":
+  version: 2.1.2
+  resolution: "graphql-modules@npm:2.1.2"
   dependencies:
     "@graphql-tools/schema": "npm:^9.0.0"
     "@graphql-tools/wrap": "npm:^9.0.0"
@@ -5466,7 +5466,7 @@ __metadata:
     ramda: "npm:^0.28.0"
   peerDependencies:
     graphql: ^14.0.0 || ^15.0.0 || ^16.0.0
-  checksum: ff6b503360bd42e4d93246a4c9cbe15d439f733fe5cbefa6bfc3f16210a6219d9a5c750371b503c6cb7ef9c7d3078b0aced34908b165fbb95adf64f09467edbb
+  checksum: aa74fd8ca0e6331150f06e71ed3ae5c0aed6d8ee39b98abe3c2f27ae498b555a111cc2295a96abfe79d042490d1747054907887da10bd99002a2085ceb0584f6
   languageName: node
   linkType: hard
 
@@ -9025,7 +9025,7 @@ __metadata:
     eslint-plugin-promise: "npm:^6.1.1"
     graphql: "npm:^16.6.0"
     graphql-import-node: "npm:^0.0.5"
-    graphql-modules: "npm:^2.1.1"
+    graphql-modules: "npm:^2.1.2"
     graphql-scalars: "npm:^1.20.4"
     jest: "npm:^29.5.0"
     prettier: "npm:^2.8.6"

niko@talos2 ~/devel/beach/server $ yarn patch -u graphql-modules
Usage Error: No package found in the project for the given locator

$ yarn patch [-u,--update] [--json] <package>

niko@talos2 ~/devel/beach/server $ ls -al .yarn/patches/
total 8
drwxr-xr-x 1 niko niko   84 Mar 23 10:27 .
drwxr-xr-x 1 niko niko   86 Mar 23 10:01 ..
-rw-r--r-- 1 niko niko 4516 Mar 23 10:27 graphql-modules-npm-2.1.1-3778b53b5f.patch
```

Any idea why?

**Reply by JavaScriptBach** [reply]
I also had a similar problem when using `yarn patch -u`. I used it to update an existing patch, with the hope that it would update my current .patch file. Instead it created a second patch file on top of the existing one, and updated the versions to strings like:

```
"mongoose": "patch:mongoose@patch%3Amongoose@npm%253A6.12.0%23./.yarn/patches/mongoose-npm-6.12.0-c2686a18c2.patch%3A%3Aversion=6.12.0&hash=c7f7cc&locator=obsidian%2540workspace%253A.#~/.yarn/patches/mongoose-patch-114d3603b5.patch"
```

```
vscode@codespaces-d51523:/workspaces/obsidian$ ls .yarn/patches/mongoose*
.yarn/patches/mongoose-npm-6.12.0-c2686a18c2.patch  .yarn/patches/mongoose-patch-114d3603b5.patch
```

I only ever want to maintain a single .patch file per library. Do we know if this is expected behavior or not? Could we consider making this feature behave more like `patch-package`? i.e. edit node_modules in place, run patch-package, patch either gets created or updated

**Reply by gozegotushar** [reply]
@JavaScriptBach just saw this locator flag in updated-patch path .... & it worked for me.
thank you man .... it took around 3 days to solve this.

**Comment by johnf** [comment]
I started a PR that add's patch: support to upgrade-interactive https://github.com/yarnpkg/berry/pull/6492

**Comment by Lakston** [comment]
I just discovered this, this is painful !

Is there really no way to (easily) apply the patch to the upgraded package ?

**Reply by darkbasic** [reply]
No there isn't. The easiest way is simply to reapply the patch from scratch.

**Reply by psychobolt** [reply]
You can use `git apply` to reapply the patch. Here are my steps: https://github.com/yarnpkg/berry/issues/4970#issuecomment-2404032671
